kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,781 @@
+
+    def __deepcopy__(self, _) -> 'Board':
+        actions = [player.next_actions for player in self.players.values()]
+        return Board(self.observation, self.configuration, actions)
+
+    def __getitem__(self, position: Point) -> Cell:
+        """
+        This method will wrap the supplied position to fit within the board size and return the cell at that location.
+        e.g. on a 3x3 board, board[(2, 1)] is the same as board[(5, 1)]
+        """
+        return self._cells[wrap_point(position, self.configuration.size)]
+
+    def __str__(self) -> str:
+        """
+        The board is printed in a grid with the following rules:
+        Capital letters are shipyards
+        Lower case letters are ships
+        Digits are cell halite and scale from 0-9 directly proportional to a value between 0 and self.configuration.max_cell_halite
+        Player 1 is letter a/A
+        Player 2 is letter b/B
+        etc.
+        """
+        size = self.configuration.size
+        result = ''
+        for y in range(size):
+            for x in range(size):
+                cell = self[(x, size - y - 1)]
+                result += '|'
+                result += (
+                    chr(ord('a') + cell.ship.player_id)
+                    if cell.ship is not None
+                    else ' '
+                )
+                normalized_halite = int(9.0 * cell.halite / float(self.configuration.max_cell_halite))
+                result += str(normalized_halite)
+                result += (
+                    chr(ord('A') + cell.shipyard.player_id)
+                    if cell.shipyard is not None
+                    else ' '
+                )
+            result += '|\n'
+        return result
+
+    def _add_ship(self: 'Board', ship: Ship) -> None:
+        ship.player.ship_ids.append(ship.id)
+        ship.cell._ship_id = ship.id
+        self.ships[ship.id] = ship
+
+    def _add_shipyard(self: 'Board', shipyard: Shipyard) -> None:
+        shipyard.player.shipyard_ids.append(shipyard.id)
+        shipyard.cell._shipyard_id = shipyard.id
+        shipyard.cell._halite = 0
+        self.shipyards[shipyard.id] = shipyard
+
+    def _delete_ship(self: 'Board', ship: Ship) -> None:
+        ship.player.ship_ids.remove(ship.id)
+        if ship.cell.ship_id == ship.id:
+            ship.cell._ship_id = None
+        del self._ships[ship.id]
+
+    def _delete_shipyard(self: 'Board', shipyard: Shipyard) -> None:
+        shipyard.player.shipyard_ids.remove(shipyard.id)
+        if shipyard.cell.shipyard_id == shipyard.id:
+            shipyard.cell._shipyard_id = None
+        del self._shipyards[shipyard.id]
+
+    def next(self) -> 'Board':
+        """
+        Returns a new board with the current board's next actions applied.
+        The current board is unmodified.
+        This can form a halite interpreter, e.g.
+            next_observation = Board(current_observation, configuration, actions).next.observation
+        """
```

Comment:
This is the replacement for the observation update logic in the halite interpreter.